src/Npgsql/PoolingDataSource.cs

```diff
@@ -116,6 +116,43 @@ class PoolingDataSource : NpgsqlDataSource
         _logger = LoggingConfiguration.ConnectionLogger;
     }
 
+    bool ReserveConnector()
+    {
+        // TODO this is a layering issue but we can't easily add this into the right place from the derived multiplexing data source.
+        // Reservations are off outside multiplexing.
+        if (!Settings.Multiplexing) return true;
+
+        NpgsqlConnector? reservedConnector = null;
+        // Try to reserve the least busy unreserved connection at most _numConnectors times to make sure we always complete.
+        for (var attempts = 0; attempts < _numConnectors; attempts++)
+        {
+            var minInFlight = int.MaxValue;
+            NpgsqlConnector? connector = null;
+            for (var i = 0; i < _numConnectors; i++)
+            {
+                var c = Connectors[i];
+                int inFlight;
+                if (connector?.ReservedForExclusiveUse == 0 && (inFlight = connector.CommandsInFlightCount) > 0)
```

Comment:
Not very important, but if we find one with CommandsInFlightCount=0 we can quite this and just try to get it via the normal path again